lib/stores/list_store.dart

```diff
@@ -69,32 +108,123 @@ class ListProvider with ChangeNotifier {
 
   Future<void> addTask({BuildContext context, Task newTask, int listId}) {
     var globalState = VikunjaGlobal.of(context);
-    _isLoading = true;
+    if (newTask.bucketId == null) _isLoading = true;
     notifyListeners();
 
     return globalState.taskService.add(listId, newTask).then((task) {
-      _tasks.insert(0, task);
+      if (_tasks.isNotEmpty)
+        _tasks.insert(0, task);
+      if (_buckets.isNotEmpty) {
+        final bucket = _buckets[_buckets.indexWhere((b) => task.bucketId == b.id)];
+        if (bucket.tasks != null) {
+          bucket.tasks.add(task);
+        } else {
+          bucket.tasks = <Task>[task];
+        }
+      }
       _isLoading = false;
       notifyListeners();
     });
   }
 
-  void updateTask({BuildContext context, int id, bool done}) {
-    var globalState = VikunjaGlobal.of(context);
-    globalState.taskService
-        .update(Task(
-      id: id,
-      done: done,
-    ))
-        .then((task) {
+  Future<Task> updateTask({BuildContext context, Task task}) {
+    return VikunjaGlobal.of(context).taskService.update(task).then((task) {
       // FIXME: This is ugly. We should use a redux to not have to do these kind of things.
       //  This is enough for now (it works™) but we should definitly fix it later.
       _tasks.asMap().forEach((i, t) {
         if (task.id == t.id) {
           _tasks[i] = task;
         }
       });
+      _buckets.asMap().forEach((i, b) => b.tasks.asMap().forEach((v, t) {
+        if (task.id == t.id){
+          _buckets[i].tasks[v] = task;
+        }
+      }));
       notifyListeners();
+      return task;
     });
   }
+
+  Future<void> addBucket({BuildContext context, Bucket newBucket, int listId}) {
+    notifyListeners();
+    return VikunjaGlobal.of(context).bucketService.add(listId, newBucket)
+        .then((bucket) {
+          _buckets.add(bucket);
+          notifyListeners();
+        });
+  }
+
+  Future<void> updateBucket({BuildContext context, Bucket bucket}) {
+    return VikunjaGlobal.of(context).bucketService.update(bucket)
+        .then((rBucket) {
+          _buckets[_buckets.indexWhere((b) => rBucket.id == b.id)] = rBucket;
+          _buckets.sort((a, b) => a.position.compareTo(b.position));
+          notifyListeners();
+        });
+  }
+  
+  Future<void> deleteBucket({BuildContext context, int listId, int bucketId}) {
+    return VikunjaGlobal.of(context).bucketService.delete(listId, bucketId)
+        .then((_) {
+          _buckets.removeWhere((bucket) => bucket.id == bucketId);
+          notifyListeners();
+        });
+  }
+
+  Future<void> moveTaskToBucket({BuildContext context, Task task, int newBucketId, int index}) async {
```

Comment:
Again, maybe worth moving the position calculation to a helper to make it testable on its own?

Reply:
Definitely a good idea.

Reply:
Done, I also made the bucket position update use it as well. `calculate_item_position.dart` is very similar to the frontend's `calculateItemPosition.ts`.